CodingConventions.md

```diff
@@ -0,0 +1,222 @@
+  func bar() err {
+     ...
+     err := foo()
+     if err != nil {
+        return fmt.Errorf("failed to do foo: %w", err)
+     }
+     ...
+  }
+  ```
+* Handle errors at a level, where you still have enough context to decide whether the error is expected during normal
+  operations.
+* Errors of unexpected types are indicators that the node's internal state might be corrupted.
+
+### anti-pattern
+
+Continuing on a best-effort basis is not an option, i.e. the following is an anti-pattern in the context of Flow:
+
+```golang
+err := foo()
+if err != nil {
+   log.Error().Err(err).Msg("foo failed")
+   return
+}
+```
+
+There are _rare_ instances, where this might be acceptable. For example, when attempting to send a message via the
+network to another node and the networking layer errored. In this case, please include a comment with an explanation,
+why it is acceptable to keep going, even if the other component returned an error.
+
+### prioritize safety over liveness
+
+**Ideally, a component should restart** (from a known good state), **when it encounters an unexpected error.**
+**If this is out of scope, we _prioritize safety over liveness_.** This means that we rather crash the node than
+continue on a best-effort basis.
+
+When in doubt, use the following as a fall-back:
+
+```golang
+err := foo()
+if err != nil {
+   if errors.Is(err, XFailedErr) {
+      // expected error
+      return
+   }
+   log.Fatal().Err(err).Msg("unexpected internal error")
+   return
+}
+```
+
+The error-handling guidelines are set up such that they minimize technical debt, even if we don't support component
+restarts right away.
+
+* The code is already differentiating expected from unexpected errors in the core business logic and we documented them
+  accordingly.
+* The code handles the expected errors at the appropriate level, only propagating unexpected errors.
+
+As a first step, you can crash the node when an unexpected error hits the `Engine` level (for more details, see `Engine`
+section of this doc). If you have followed the error handling guidelines, leaving the restart logic for the component as
+technical debt for later hopefully does not add much overhead. In contrast, if you log errors and continue on a
+best-effort basis, you also leave the _entire_ logic for differentiating errors as tech debt. When adding this logic
+later, you need to revisit the _entire_ business logic again.
+
+## Engines
+
+On a high level, a Flow node consists of various data-processing components that receive information from node-internal
+and external sources. We model the data flow inside a Flow node as a graph:
+a data-processing component forms a vertex; two components exchanging messages is represented as an edge. An `Engine` is
+the interface for a data-processing component.
+
+Generally, we consider node-internal components as _trusted_ and external nodes as untrusted data sources. The `Engine`
+API differentiates between trusted and untrusted inputs:
```

Comment:
Given that the engine API is deprecated in favour of `MessageProcessor`, maybe we could use different examples here. 

I created a PR adding deprecation flags with some suggested alternatives to the deprecated Engine methods: https://github.com/onflow/flow-go/pull/2600. If we have consensus on those suggested alternative, I can help update the examples here as well.

Reply:
Alex added a section here with a different example demonstrating this pattern: https://github.com/onflow/flow-go/pull/2575/files#diff-3b52addb04dfea0ae48593e2aea4efa434d108d00bc9df92f8b51f170dc807fbR41-R48